### PR TITLE
Exclude admin / diagnostic requests from HTTP request limiting

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we exclude certain HTTP requests that are needed to inspect the cluster
from HTTP request limiting to ensure these commands are processed even in critical
memory conditions.

Relates #17951, relates #18145

### CODE REVIEW

**Comment by s1monw** [comment]
I wonder if we should go from a whitelist to a blacklist, it seems like we are adding more and more exceptions - maybe we should just put the memory hungry ones in there explicitly?

**Comment by danielmitterdorfer** [comment]
Previous exceptions were defined on transport level (OTOH, we have currently 3 exceptions defined on transport level). This PR adds support for HTTP level. I think the better option is to be conservative with exceptions and just a whitelist but it would be pretty easy to use a blacklist instead (just swap the default and blacklist the relevant actions).

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 1]
```diff
@@ -290,4 +301,22 @@ public void process(RestRequest request, RestChannel channel, RestFilterChain fi
             executeHandler(request, channel);
         }
     }
+
+    private static final class RestHandlerHolder {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: would it make more sense to add a method to the `RestHandler` interface `default boolean canTripCircuitBreaker() { return true; }` that would safe us all the extra indirection and is simple to override?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: it would also be consistent across all the methods the handler is registered?

**Comment by s1monw** [comment]
> Previous exceptions were defined on transport level (OTOH, we have currently 3 exceptions defined on transport level). This PR adds support for HTTP level. I think the better option is to be conservative with exceptions and just use a whitelist but it would be pretty easy to use a blacklist instead (just swap the default and blacklist the relevant actions).

ok sure.... I left some comments

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 1 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I also considered this but did not want to mix various aspects in the `RestHandler` interface. But I like the approach and can change that.

**Comment by danielmitterdorfer** [comment]
@s1monw I've pushed a commit that addresses your comments. Could you please have another look?

**Comment by s1monw** [comment]
LGTM

**Comment by danielmitterdorfer** [comment]
Thanks; I'll merge soon.

**Pull request closed without merging**